### PR TITLE
Issues # 7627

### PR DESCRIPTION
**Pull request by Celthi**
Fixed the Issue #7627. According the download Json data, changed the video url construct a bit.

### CODE REVIEW

File: youtube_dl/extractor/youku.py [hunk 1]
```diff
@@ -100,15 +101,24 @@ def yk_t(s1, s2):
...
-                int(i) for i in data1['streamfileids'][format].strip('*').split('*')]
-            fileid = ''.join(
-                [shuffled_string_ls[i] for i in streamfileid])
-            fileid_dict[format] = fileid[:8] + '%s' + fileid[10:]
+        for stream in data1['stream']:
+            format = stream.get('stream_type')
+            #streamfileid = [
+            #    int(i) for i in data1['stream']['streamfileids']#[format].strip('*').split('*')]
+            #fileid = ''.join(
+            #    [shuffled_string_ls[i] for i in streamfileid])
```
Comment (yan12125) [inline youtube_dl/extractor/youku.py]: Please remove them. We don't keep unused codes in the codebase.

File: youtube_dl/extractor/youku.py [hunk 2]
```diff
@@ -208,7 +234,8 @@ def retrieve_data(req_url, note):
             basic_data_url,
             'Downloading JSON metadata 1')
         data2 = retrieve_data(
-            'http://v.youku.com/player/getPlayList/VideoIDS/%s/Pf/4/ctype/12/ev/1' % video_id,
+            #'http://v.youku.com/player/getPlayList/VideoIDS/%s/Pf/4/ctype/12/ev/1' % video_id,
+            "http://play.youku.com/play/get.json?vid=%s&ct=12" % video_id,
```
Comment (yan12125) [inline youtube_dl/extractor/youku.py]: ```data1``` and ```data2``` use the same URL. Please merge their usages.

**Comment by yan12125** [comment]
Did you borrow codes from you-get? If so, you need to ask for explicit permission to use their codes in youtube-dl from its authors.

**Comment by yan12125** [comment]
Fails for test_Youku_2:
```
$ youtube-dl -v "http://v.youku.com/v_show/id_XMTI1OTczNDM5Mg==.html"
[debug] System config: []
[debug] User config: []
[debug] Command-line args: ['-v', 'http://v.youku.com/v_show/id_XMTI1OTczNDM5Mg==.html']
[debug] Encodings: locale UTF-8, fs utf-8, out UTF-8, pref UTF-8
[debug] youtube-dl version 2015.12.10
[debug] Git HEAD: b6afc22
[debug] Python version 3.5.1 - Linux-4.3.0-1-ARCH-x86_64-with-arch-Arch-Linux
[debug] exe versions: ffmpeg 2.8.3, ffprobe 2.8.3, rtmpdump 2.4
[debug] Proxy map: {}
[youku] XMTI1OTczNDM5Mg: Downloading JSON metadata 1
[youku] XMTI1OTczNDM5Mg: Downloading JSON metadata 2
ERROR: An extractor error has occured. (caused by KeyError('mp4hd',)); please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; see  https://yt-dl.org/update  on how to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
Traceback (most recent call last):
  File "/home/yen/Executables/Multimedia/youtube-dl/youtube_dl/extractor/common.py", line 290, in extract
    return self._real_extract(url)
  File "/home/yen/Executables/Multimedia/youtube-dl/youtube_dl/extractor/youku.py", line 257, in _real_extract
    video_urls_dict = self.construct_video_urls(data1, data2)
  File "/home/yen/Executables/Multimedia/youtube-dl/youtube_dl/extractor/youku.py", line 144, in construct_video_urls
    'hd': self.get_hd(format),
  File "/home/yen/Executables/Multimedia/youtube-dl/youtube_dl/extractor/youku.py", line 176, in get_hd
    return hd_id_dict[fm]
KeyError: 'mp4hd'
Traceback (most recent call last):
  File "/home/yen/Executables/Multimedia/youtube-dl/youtube_dl/extractor/common.py", line 290, in extract
    return self._real_extract(url)
  File "/home/yen/Executables/Multimedia/youtube-dl/youtube_dl/extractor/youku.py", line 257, in _real_extract
    video_urls_dict = self.construct_video_urls(data1, data2)
  File "/home/yen/Executables/Multimedia/youtube-dl/youtube_dl/extractor/youku.py", line 144, in construct_video_urls
    'hd': self.get_hd(format),
  File "/home/yen/Executables/Multimedia/youtube-dl/youtube_dl/extractor/youku.py", line 176, in get_hd
    return hd_id_dict[fm]
KeyError: 'mp4hd'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/yen/Executables/Multimedia/youtube-dl/youtube_dl/YoutubeDL.py", line 663, in extract_info
    ie_result = ie.extract(url)
  File "/home/yen/Executables/Multimedia/youtube-dl/youtube_dl/extractor/common.py", line 296, in extract
    raise ExtractorError('An extractor error has occured.', cause=e)
youtube_dl.utils.ExtractorError: An extractor error has occured. (caused by KeyError('mp4hd',)); please report this issue on https://yt-dl.org/bug . Make sure you are using the latest version; see  https://yt-dl.org/update  on how to update. Be sure to call youtube-dl with the --verbose flag and include its complete output.
```

File: youtube_dl/extractor/youku.py (hunk 2 above)
Comment (Celthi) [inline youtube_dl/extractor/youku.py]: when I cloned the codes, data1 and data2 are already there. I try to not rewrite the code , instead fix in the original codes. So I keep data1 and data2.

File: youtube_dl/extractor/youku.py (hunk 1 above)
Comment (Celthi) [inline youtube_dl/extractor/youku.py]: Thank you for the reminding. This is my first commit to a open source. I will try my best.

**Comment by Celthi** [comment]
The KeyError('mp4hd') has fixed, and it work fine except the video protected by the password.

File: youtube_dl/extractor/youku.py (hunk 2 above)
Comment (yan12125) [inline youtube_dl/extractor/youku.py]: I know that. If Youku merges the two API calls, we should do the same thing.

**Comment by yan12125** [comment]
Thanks merged. Would you like to be listed in [AUTHORS](https://github.com/rg3/youtube-dl/blob/master/AUTHORS) under the name Evan Lu?

**Comment by Celthi** [comment]
No Problem.  I'm glad I can contribute to this fabulous project.

**Comment by yan12125** [comment]
Done in 174964a7bc4b8d759a4019e366520c9121fcbf15.

**Comment by Celthi** [comment]
Thank you!

**Comment by yan12125** [comment]
Well I found that ```mp4hd3``` exists in [```parse_ext_l```](https://github.com/rg3/youtube-dl/blob/174964a7bc4b8d759a4019e366520c9121fcbf15/youtube_dl/extractor/youku.py#L167) and [```get_format_name```](https://github.com/rg3/youtube-dl/blob/174964a7bc4b8d759a4019e366520c9121fcbf15/youtube_dl/extractor/youku.py#L182) but not in [```get_hd```](https://github.com/rg3/youtube-dl/blob/174964a7bc4b8d759a4019e366520c9121fcbf15/youtube_dl/extractor/youku.py#L144-L156). Which value should it be?

**Comment by Celthi** [comment]
It should be '1'. Sorry for this.

**Comment by yan12125** [comment]
Thanks fixed in deb1e8d20e8680d8e217e2d376e05836541a46ca. Everything is done! The next time please:
1. Prefix commit messages with the extractor name. For example: ```[youku] merge data1 and data2```
2. It's recommended (not mandatory) to check codes with [flake8](https://github.com/rg3/youtube-dl#adding-support-for-a-new-site) before committing.

**Comment by Celthi** [comment]
No problem. Next time I will check with flake8.

**Pull request merged**